airflow/models/serialized_dag.py

```diff
@@ -94,13 +97,26 @@ def write_dag(cls, dag: DAG, min_update_interval: Optional[int] = None, session=
                      (timezone.utcnow() - timedelta(seconds=min_update_interval)) < cls.last_updated))
             ).scalar():
                 return
-        log.debug("Writing DAG: %s to the DB", dag.dag_id)
-        session.merge(cls(dag))
+
+        log.debug("Checking if DAG (%s) is updated or same", dag.dag_id)
+        serialized_dag_from_db: SerializedDagModel = (
+            session
+            .query(cls)
+            .filter(cls.dag_id == dag.dag_id)
+            .one_or_none()
+        )
```

Comment:
```suggestion
        serialized_dag_from_db: SerializedDagModel = session.query(cls).get(dag.dag_id)
```

Reply:
`get()` docs https://docs.sqlalchemy.org/en/13/orm/query.html?highlight=query%20get#sqlalchemy.orm.query.Query.get

Reply:
`get` will only provide performance boost if the object is in memory isn't it?

Otherwise under the hood, get calls `one`  which in turn calls `one_or_none`:

https://github.com/sqlalchemy/sqlalchemy/blob/f4aa9def8f35136d1fe3bba81aaeddb72e7772a2/lib/sqlalchemy/orm/query.py#L907

https://github.com/sqlalchemy/sqlalchemy/blob/rel_1_3_17/lib/sqlalchemy/orm/loading.py#L287

https://github.com/sqlalchemy/sqlalchemy/blob/f4aa9def8f35136d1fe3bba81aaeddb72e7772a2/lib/sqlalchemy/orm/query.py#L2641-L2659

Reply:
Given we want to check that the SerializedDAG is changed or not, we want to access the DB rather than getting anything from the memory.

I feel `.one_or_none` is better here but happy to hear your thoughts.

Reply:
In the AirBnB talk they mentioned about really big DAGs and the need to use compression (they implemented it in their talk.  I think adding one extra read here might add quite heavy DB throughput. It's an edge case, but possibly an important one and some companies already experiencing it.

I know it would be a bit bigger change (involving DB migration) but possibly we should calculate hash of the DAG and compare it and read it from the DB rather than then DAG itself. That seems like perfectly doable and not that complex.

Reply:
Yeah already going to do that one in other PR. Calculating hash of the serialized JSON. But want to do it in a separate PR

Reply:
Cool!

Reply:
In this case query is better as you will have to replace it with https://docs.sqlalchemy.org/en/13/orm/query.html#sqlalchemy.orm.query.Query.scalar later.

Reply:
Updated in https://github.com/apache/airflow/pull/9850/commits/453dee82276abae75488a4434968da187ccce0a3